File: solutions/Climbing-Stairs/Climbing-Stairs.cpp

// class Solution {
// public:
//     int climbStairs(int n) {
//         if(n==1)
//             return 1;
//         int a=1,b=1,ans=0;
//         for(int i=2;i<=n;i++)
//         {
//             ans=a+b;
//             a=b;
//             b=ans;
//         }
//         return ans;
//     }
// };

// 2019-07-21
// O(n)
class Solution {
public:
    int climbStairs(int n) {
        if(n==0)
        {
            return 0;
        }
        if (n==1)
        {
            return 1;
        }
        if(n==2)
        {
            return 2;
        }
        
        int slow = 1;
        int fast = 2;
        int ans = 0;
        for(int i=3;i<=n;i++)
        {
            ans = slow + fast;
            slow = fast;
            fast = ans;
        }
        return ans;
    }
};
 

File: solutions/Climbing-Stairs/Climbing-Stairs.py


class Solution:
    # @param n, an integer
    # @return an integer
    def climbStairs(self, n):
        dp=range(0,1000)
        dp[1]=1
        dp[2]=2
        for i in range(3,n+1):
            dp[i]=dp[i-1]+dp[i-2]
        return dp[n]
 